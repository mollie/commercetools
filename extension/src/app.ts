import express, { Application } from 'express';
import handler from '../expressHandler';
import morganMiddleware from './logger/morganMiddleware';

const app: Application = express();

app.use(morganMiddleware); // Logs from http level

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use(express.json());

app.use('/', handler);

export default app;
