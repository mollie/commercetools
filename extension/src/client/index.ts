import createMollieClient, { MollieClient } from '@mollie/api-client';
import { ClientBuilder, Client, AuthMiddlewareOptions, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}

export function initialiseCommercetoolsClient(): Client {
  const {
    commercetools: { projectKey, clientId, clientSecret, host, authUrl, scopes },
  } = config;

  const authMiddlewareOptions: AuthMiddlewareOptions = {
    host: authUrl,
    projectKey,
    credentials: {
      clientId,
      clientSecret,
    },
    scopes,
    fetch,
  };

  const httpMiddlewareOptions: HttpMiddlewareOptions = {
    host,
    fetch,
  };

  const client: Client = new ClientBuilder()
    .withClientCredentialsFlow(authMiddlewareOptions)
    .withHttpMiddleware(httpMiddlewareOptions)
    .withUserAgentMiddleware()
    .build();

  return client;
}
