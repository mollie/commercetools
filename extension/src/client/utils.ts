import createMollieClient, { MollieClient } from '@mollie/api-client';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}
