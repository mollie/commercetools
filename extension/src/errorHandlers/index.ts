import { formatMollieErrorResponse } from './formatMollieErrorResponse';
import { formatExtensionErrorResponse } from './formatExtensionErrorResponse';

const errorHandler = {
  mollie: formatMollieErrorResponse,
  extension: formatExtensionErrorResponse,
};

export default errorHandler;
