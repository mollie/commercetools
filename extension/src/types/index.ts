// This is not exhaustive
// If you use another commercetools error response code, add it to this enum
export enum CTEnumErrors {
  General = 'General',
  InvalidInput = 'InvalidInput',
  InvalidOperation = 'InvalidOperation',
  Unauthorized = 'Unauthorized',
  SyntaxError = 'SyntaxError',
  SemanticError = 'SemanticError',
  ObjectNotFound = 'ObjectNotFound',
}

export type CTUpdatesRequestedResponse = {
  status: number;
  actions?: Action[];
  errors?: CTError[];
};

export type Action = {
  action: string;
  type?: {
    key: string;
  };
  fields?: {
    actionType: string;
    request?: string;
    response?: string;
    createdAt?: string;
  };
  name?: string;
  value?: string;
  key?: string;
  transaction?: CTTransaction;
  transactionId?: string;
  interactionId?: string;
  state?: CTTransactionState;
};

export type CTError = {
  code: CTEnumErrors;
  message: string;
  extensionExtraInfo?: Object;
};

export type CTMoney = {
  type?: 'centPrecision';
  currencyCode: string;
  centAmount: number;
  fractionDigits?: number;
};

export type CTPayment = {
  id: string;
  amountPlanned: CTMoney;
  transactions?: CTTransaction[];
  key?: string;
  custom?: {
    fields?: {
      paymentMethodsRequest?: string;
      paymentMethodsResponse?: string;
      createPayment?: string;
      createCapture?: string;
      cancelPayment?: string;
      createRefund?: string;
    };
  };
  interfaceInteractions?: CTInterfaceInteraction[];
};

export type CTInterfaceInteraction = {
  id: string;
  actionType: string;
  createdAt?: Date;
  request?: string;
  response?: string;
};

export type CTTransaction = {
  id?: string;
  timestamp?: string;
  type: CTTransactionType;
  amount: CTMoney;
  interactionId?: string;
  state?: 'Initial' | 'Pending' | 'Success' | 'Failure';
};

export enum ControllerAction {
  GetPaymentMethods = 'getPaymentMethods',
  CreateOrder = 'createOrder',
  CreateOrderPayment = 'createOrderPayment',
  CreateOrderRefund = 'createOrderRefund',
  CreateCustomRefund = 'createCustomRefund',
  CreateShipment = 'createShipment',
  UpdateShipment = 'updateShipment',
  CancelOrder = 'cancelOrder',
  NoAction = 'noAction',
  Error = 'error',
}

export enum CTTransactionType {
  Authorization = 'Authorization',
  CancelAuthorization = 'CancelAuthorization',
  Charge = 'Charge',
  Refund = 'Refund',
  Chargeback = 'Chargeback',
}

export enum CTTransactionState {
  Initial = 'Initial',
  Pending = 'Pending',
  Success = 'Success',
  Failure = 'Failure',
}
