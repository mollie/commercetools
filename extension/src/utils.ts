import { MethodsListParams } from '@mollie/api-client';
/**
 * Generates an ISO string date
 * @returns {String} Returns the current date converted to ISO.
 */
export function createDateNowString() {
  return new Date().toISOString();
}

export function amountMapper(amountPlanned: any): string {
  const { centAmount, fractionDigits } = amountPlanned;
  const divider = Math.pow(10, fractionDigits || 2);
  const mollieAmount = (centAmount / divider).toFixed(2);
  return mollieAmount
}

export function methodListMapper(ctObj: any): MethodsListParams {
  // Generally this shouldn't be needed, but a safety anyway.. eventually could return error here
  if (!ctObj.amountPlanned) {
    return {};
  }
  const mollieAmount = amountMapper(ctObj.amountPlanned);

  const mObject: MethodsListParams = {
    amount: {
      value: mollieAmount,
      currency: ctObj.amountPlanned.currencyCode,
    },
    // Resource is hardcoded, for the time being we only support Orders API
    resource: 'orders',
  };

  if (ctObj.custom?.fields?.paymentMethodsRequest) {
    const parsedMethodsRequest = JSON.parse(ctObj.custom?.fields?.paymentMethodsRequest)
    const { locale, billingCountry, includeWallets, orderLineCategories, issuers, pricing, sequenceType } = parsedMethodsRequest;
    const include = issuers || pricing ? `${issuers ? 'issuers,' : ''}${pricing ? 'pricing' : ''}` : undefined;

    Object.assign(
      mObject,
      locale && { locale: locale },
      include && { include: include },
      includeWallets && { includeWallets: includeWallets },
      billingCountry && { billingCountry: billingCountry },
      sequenceType && { sequenceType: sequenceType },
      orderLineCategories && { orderLineCategories: orderLineCategories },
    );
  }

  return mObject;
}
