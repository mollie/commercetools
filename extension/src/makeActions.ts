import { time } from 'console';
import { v4 as uuid } from 'uuid';
import { Action, ControllerAction, CTTransactionState } from './types';
import { createDateNowString } from './utils';

/**
 *
 * @param customFieldName
 * @param customFieldValue
 * If the customFieldValue is an API response, JSON Stringify it before passing it
 */
const setCustomField = (customFieldName: string, customFieldValue: string) => {
  return {
    action: 'setCustomField',
    name: customFieldName,
    value: customFieldValue,
  };
};

type createInterfaceInteractionParams = {
  actionType: ControllerAction;
  requestValue: string;
  responseValue: string;
  id?: string;
  timestamp?: string;
};
/**
 * @param parameters type createInterfaceInteractionParams, which contains
 * actionType ControllerAction
 * requestValue string
 * responseValue string
 * id string string, optional
 * timestamp string, optional
 * If the responseValue is an API response, JSON Stringify it before passing it
 */
const addInterfaceInteraction = (params: createInterfaceInteractionParams) => {
  const { actionType, requestValue, responseValue, id, timestamp } = params;
  const interfaceInteractionId = id ? id : uuid();
  const interfaceInteractionTimestamp = timestamp ? timestamp : createDateNowString();
  return {
    action: 'addInterfaceInteraction',
    type: {
      key: 'ct-mollie-integration-interface-interaction-type',
    },
    fields: {
      id: interfaceInteractionId,
      actionType,
      createdAt: interfaceInteractionTimestamp,
      request: requestValue,
      response: responseValue,
    },
  };
};

/**
 *
 * @param key will be the mollie order id
 */
const setKey = (key: string) => {
  return {
    action: 'setKey',
    key,
  };
};

/**
 *
 * @param interfaceText will be the mollie order status
 */
const setStatusInterfaceText = (interfaceText: string) => {
  return {
    action: 'setMethodInfoName',
    interfaceText,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param newState CTTransactionState
 */
const changeTransactionState = (id: string, newState: CTTransactionState) => {
  return {
    action: 'changeTransactionState',
    transactionId: id,
    state: newState,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param interactionId either the mollie payment id, or the corresponding interfaceInteraction's id
 */
const changeTransactionInteractionId = (id: string, interactionId: string) => {
  return {
    action: 'changeTransactionInteractionId',
    transactionId: id,
    interactionId,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param timestamp CT DateTime is a JSON string representation of UTC date & time in ISO 8601 format (YYYY-MM-DDThh:mm:ss.sssZ)
 * for example: "2018-10-12T14:00:00.000Z"
 */
const changeTransactionTimestamp = (id: string, timestamp: string) => {
  return {
    action: 'changeTransactionTimestamp',
    transactionId: id,
    timestamp,
  };
};

export const makeActions = {
  setCustomField,
  addInterfaceInteraction,
  setKey,
  setStatusInterfaceText,
  changeTransactionState,
  changeTransactionInteractionId,
  changeTransactionTimestamp,
};
