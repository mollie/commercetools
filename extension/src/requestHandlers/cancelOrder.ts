import { MollieClient, Order, OrderLineCancelParams } from '@mollie/api-client';
import Logger from '../logger/logger';
import { formatMollieErrorResponse } from '../errorHandlers/formatMollieErrorResponse';
import { Action, ControllerAction, CTUpdatesRequestedResponse } from '../types';
import { createDateNowString } from '../utils';

export function createCtActions(mollieCancelOrderRes: Order, ctObj: any): Action[] {
  const stringifiedCancelOrderResponse = JSON.stringify(mollieCancelOrderRes);
  const result: Action[] = [
    {
      action: 'addInterfaceInteraction',
      type: {
        key: 'ct-mollie-integration-interface-interaction-type',
      },
      fields: {
        actionType: ControllerAction.CancelOrder,
        createdAt: createDateNowString(),
        request: ctObj?.custom?.fields?.cancelOrderRequest,
        response: stringifiedCancelOrderResponse,
      },
    },
    {
      action: 'setCustomField',
      name: 'createCancelOrderResponse',
      value: stringifiedCancelOrderResponse,
    },
  ];
  return result;
}

export default async function cancelOrder(ctObj: any, mollieClient: MollieClient, createCtActions: Function): Promise<CTUpdatesRequestedResponse> {
  try {
    const mollieCancelOrderRes = await mollieClient.orders.cancel(ctObj.key);
    Logger.debug(mollieCancelOrderRes);
    const ctActions = createCtActions(mollieCancelOrderRes, ctObj);
    return {
      actions: ctActions,
      status: 200,
    };
  } catch (error: any) {
    Logger.error(error);
    const errorResponse = formatMollieErrorResponse(error);
    return errorResponse;
  }
}
