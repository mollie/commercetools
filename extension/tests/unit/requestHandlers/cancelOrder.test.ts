import { v4 as uuid } from 'uuid';
import { mocked } from 'ts-jest/utils';
import { Action, ControllerAction } from '../../../src/types';
import cancelOrder, { getCancelOrderParams, createCtActions } from '../../../src/requestHandlers/cancelOrder';
import { createDateNowString, makeMollieLineAmounts } from '../../../src/utils';
import { makeActions } from '../../../src/makeActions';
import Logger from '../../../src/logger/logger';

jest.mock('uuid');
jest.mock('../../../src/utils');
jest.mock('../../../src/makeActions');

describe('getCancelOrderParams', () => {
  const mockLogError = jest.fn();
  const mockUuid = '238e8459-06a1-46f6-95c6-cf3ce0998dce';
  beforeEach(() => {
    mocked(uuid).mockReturnValue(mockUuid);
    Logger.error = mockLogError;
    mocked(makeMollieLineAmounts).mockReturnValueOnce([{ id: 'odl_1.n3xdt3', quantity: 1, amount: { currency: 'EUR', value: '4.20' } }]);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should create required params for mollie cancelOrder call', async () => {
    const mockedCtObj = {
      key: 'ord_3uwvfd',
      custom: {
        fields: {
          createCancelOrderRequest: '[{"id":"odl_1.n3xdt3","quantity":1,"amount":{"currencyCode":"EUR","centAmount":420,"fractionDigits":2}}]',
        },
      },
    };
    const expectedCreateCancelOrderParams = {
      orderId: 'ord_3uwvfd',
      lines: [{ id: 'odl_1.n3xdt3', quantity: 1, amount: { currency: 'EUR', value: '4.20' } }],
    };

    await expect(getCancelOrderParams(mockedCtObj)).resolves.toEqual(expectedCreateCancelOrderParams);
  });

  it('Should return 400 and error message if creating the parameters throws an error', async () => {
    const mockedCtObj = {
      key: 'ord_3uwvfd',
      custom: {
        fields: {
          createCancelOrderRequest: 'unparsableJson',
        },
      },
    };
    const expectedRejectedValue = {
      status: 400,
      title: 'Could not make parameters required to cancel Mollie order.',
      field: 'createCancelOrderRequest',
    };
    await expect(getCancelOrderParams(mockedCtObj)).rejects.toEqual(expectedRejectedValue);
    expect(mockLogError).toHaveBeenCalledTimes(1);
  });
});

describe('createCtActions', () => {
  beforeEach(() => {
    mocked(createDateNowString).mockReturnValue('2021-10-08T12:12:02.625Z');
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should create correct ct actions from request and mollies response', () => {
    const mockedCtObject = {
      custom: {
        fields: {
          createCancelOrderRequest: '[]',
        },
      },
    };

    const mockedCancelOrderResponse: any = {
      resource: 'order',
      id: 'ord_jwtj05',
      profileId: 'pfl_VtWA783A63',
      amount: { value: '20.00', currency: 'EUR' },
      status: 'canceled',
      isCancelable: false,
      createdAt: '2021-11-01T12:20:39+00:00',
      canceledAt: '2021-11-02T12:26:44+00:00',
      redirectUrl: 'https://www.mollie.com/',
      lines: [
        {
          resource: 'orderline',
          id: 'odl_1.694ky5',
          orderId: 'ord_jwtj05',
          name: 'apple',
          sku: null,
          type: 'physical',
          status: 'canceled',
          isCancelable: false,
          quantity: 1,
          quantityShipped: 0,
          quantityCanceled: 1,
          amountCanceled: {},
          createdAt: '2021-11-01T12:20:39+00:00',
        },
      ],
    };
    mocked(makeActions.addInterfaceInteraction).mockReturnValue({
      action: 'addInterfaceInteraction',
      type: {
        key: 'ct-mollie-integration-interface-interaction-type',
      },
      fields: {
        id: uuid(),
        actionType: ControllerAction.CancelOrder,
        createdAt: '2021-10-08T12:12:02.625Z',
        request: '[]',
        response: JSON.stringify(mockedCancelOrderResponse),
      },
    });
    mocked(makeActions.setCustomField).mockReturnValue({
      action: 'setCustomField',
      name: 'createCancelOrderResponse',
      value: JSON.stringify(mockedCancelOrderResponse),
    });
    const ctActions = createCtActions(mockedCancelOrderResponse, mockedCtObject);
    ctActions.forEach(action => {
      expect(action).toMatchSnapshot();
    });
  });
});

describe('cancelOrder', () => {
  const mockLoggerError = jest.fn();
  beforeEach(() => {
    Logger.error = mockLoggerError;
    mocked(createDateNowString).mockReturnValue('2021-10-08T12:12:02.625Z');
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should call mollie, handle response and return actions when cancelling complete order', async () => {
    const mockedCtObject = {
      key: 'ord_jwtj05',
      custom: {
        fields: {
          createCancelOrderRequest: '[]',
        },
      },
    };
    const mockedCancelOrderResponse: any = {
      resource: 'order',
      id: 'ord_jwtj05',
      profileId: 'pfl_VtWA783A63',
      amount: { value: '20.00', currency: 'EUR' },
      status: 'canceled',
      isCancelable: false,
      createdAt: '2021-11-01T12:20:39+00:00',
      canceledAt: '2021-11-02T12:26:44+00:00',
      redirectUrl: 'https://www.mollie.com/',
      lines: [
        {
          resource: 'orderline',
          id: 'odl_1.694ky5',
          orderId: 'ord_jwtj05',
          name: 'apple',
          sku: null,
          type: 'physical',
          status: 'canceled',
          isCancelable: false,
          quantity: 1,
          quantityShipped: 0,
          quantityCanceled: 1,
          amountCanceled: {},
          createdAt: '2021-11-01T12:20:39+00:00',
        },
      ],
    };
    const mockedCtActions: Action[] = [];
    const getCancelOrderParams = jest.fn().mockResolvedValueOnce({ orderId: 'ord_jwtj05', lines: [] });
    const createCtActions = jest.fn().mockReturnValueOnce(mockedCtActions);
    const mollieClient = { orders: { cancel: jest.fn().mockResolvedValueOnce(mockedCancelOrderResponse) } } as any;

    const cancelOrderRes = await cancelOrder(mockedCtObject, mollieClient, getCancelOrderParams, createCtActions);
    expect(mollieClient.orders.cancel).toHaveBeenCalledWith(mockedCtObject.key);
    expect(createCtActions).toBeCalledWith(mockedCancelOrderResponse, mockedCtObject);
    expect(cancelOrderRes.status).toBe(200);
  });
  it('Should call mollie, handle response and return actions when cancelling partial order', async () => {
    const mockedCtObject = {
      key: 'ord_jwtj05',
      custom: {
        fields: {
          createCancelOrderRequest: '[{"id":"odl_1.694ky5","quantity":1,"amount":{"currencyCode":"EUR","centAmount":1430,"fractionDigits":2}}]',
        },
      },
    };
    const mockedCancelOrderParams = {
      orderId: 'ord_jwtj05',
      lines: [{ id: 'odl_1.694ky5', quantity: 1, amount: { currency: 'EUR', value: '14.30' } }],
    };
    const mockedCancelOrderResponse: any = {
      resource: 'order',
      id: 'ord_jwtj05',
      profileId: 'pfl_VtWA783A63',
      amount: { value: '14.30', currency: 'EUR' },
      status: 'canceled',
      isCancelable: false,
      createdAt: '2021-11-01T12:20:39+00:00',
      canceledAt: '2021-11-02T12:26:44+00:00',
      lines: [
        {
          resource: 'orderline',
          id: 'odl_1.694ky5',
          orderId: 'ord_jwtj05',
          name: 'apple',
          sku: null,
          type: 'physical',
          status: 'canceled',
          isCancelable: false,
          quantity: 1,
          quantityShipped: 0,
          quantityCanceled: 1,
          amountCanceled: {},
          createdAt: '2021-11-01T12:20:39+00:00',
        },
      ],
    };
    const mockedCtActions: Action[] = [];
    const getCancelOrderParams = jest.fn().mockResolvedValueOnce(mockedCancelOrderParams);
    const createCtActions = jest.fn().mockReturnValueOnce(mockedCtActions);
    const mollieClient = { orders_lines: { cancel: jest.fn().mockResolvedValueOnce(mockedCancelOrderResponse) } } as any;

    const cancelOrderRes = await cancelOrder(mockedCtObject, mollieClient, getCancelOrderParams, createCtActions);
    expect(mollieClient.orders_lines.cancel).toHaveBeenCalledWith(mockedCancelOrderParams);
    expect(createCtActions).toBeCalledWith(mockedCancelOrderResponse, mockedCtObject);
    expect(cancelOrderRes.status).toBe(200);
  });
  it('Should return commercetools formated error if one of the functions fails', async () => {
    const mockedError = { status: 400, title: 'The order cannot be canceled from state: canceled', field: 'createCancelOrderRequest' };
    const createCtActions = jest.fn();
    const getCancelOrderParams = jest.fn();
    const mollieClient = { orders: { cancel: jest.fn().mockRejectedValueOnce(mockedError) } } as any;

    const cancelOrderRes = await cancelOrder({}, mollieClient, getCancelOrderParams, createCtActions);
    expect(cancelOrderRes.status).toBe(400);
    expect(cancelOrderRes.errors).toHaveLength(1);
  });
});
