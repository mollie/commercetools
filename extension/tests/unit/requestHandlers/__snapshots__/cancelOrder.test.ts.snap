// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCtActions Should create correct ct actions from request and mollies payment for order lines (mollie response is true) 1`] = `
{
  "action": "changeTransactionState",
  "state": "Success",
  "transactionId": "869ea4f0-b9f6-4006-bf04-d8306b5c9564",
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment for order lines (mollie response is true) 2`] = `
{
  "action": "changeTransactionTimestamp",
  "timestamp": "2021-10-08T12:12:02.625Z",
  "transactionId": "869ea4f0-b9f6-4006-bf04-d8306b5c9564",
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment for order lines (mollie response is true) 3`] = `
{
  "action": "changeTransactionInteractionId",
  "interactionId": "3fea7470-5434-4056-a829-a187339e94d8",
  "transactionId": "869ea4f0-b9f6-4006-bf04-d8306b5c9564",
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment for order lines (mollie response is true) 4`] = `
{
  "action": "addInterfaceInteraction",
  "fields": {
    "actionType": "cancelOrder",
    "createdAt": "2021-10-08T12:12:02.625Z",
    "id": "3fea7470-5434-4056-a829-a187339e94d8",
    "request": "{"transactionId":"869ea4f0-b9f6-4006-bf04-d8306b5c9564","cancelOrder":{"lineIds":"3e632c95-8dc6-459a-9edc-5e64760abf21,  31fb05cd-6e00-42b4-a25d-68cceef5a603","includeShipping":true}}",
    "response": "Ok",
  },
  "type": {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment for whole cancel order (mollie response is object) 1`] = `
{
  "action": "changeTransactionState",
  "state": "Success",
  "transactionId": "869ea4f0-b9f6-4006-bf04-d8306b5c9564",
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment for whole cancel order (mollie response is object) 2`] = `
{
  "action": "changeTransactionTimestamp",
  "timestamp": "2021-10-08T12:12:02.625Z",
  "transactionId": "869ea4f0-b9f6-4006-bf04-d8306b5c9564",
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment for whole cancel order (mollie response is object) 3`] = `
{
  "action": "changeTransactionInteractionId",
  "interactionId": "3fea7470-5434-4056-a829-a187339e94d8",
  "transactionId": "869ea4f0-b9f6-4006-bf04-d8306b5c9564",
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment for whole cancel order (mollie response is object) 4`] = `
{
  "action": "addInterfaceInteraction",
  "fields": {
    "actionType": "cancelOrder",
    "createdAt": "2021-10-08T12:12:02.625Z",
    "id": "3fea7470-5434-4056-a829-a187339e94d8",
    "request": "{"transactionId":"869ea4f0-b9f6-4006-bf04-d8306b5c9564","cancelOrder":{}}",
    "response": "{"lineIds":"42c6d1fd-b942-433b-b6dd-41062c4b3a42"}",
  },
  "type": {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;
